### PR TITLE
ci(editor): Run e2e tests in parallel and improve build caching

### PR DESCRIPTION
**Pull request by OlegIvaniv**
With the increasing amount of E2E specs, our testing pipeline would often crash, or time out after 30 minutes. To tackle this issue, we'll run the E2E specs in parallel. This PR contains two main parts:

## 1. Make the E2E Workflow reusable 
The E2E workflow is now reusable and accepts several input parameters which allow the specification of: 
- branch
- optional run environment(not just node versions, but also different browsers)
- optional spec matcher
- optional user
- optional container matrix
- optional parallel toggle
- optional record toggle

We can use this workflow(`workflows/e2e-reusable.yml`) in the parent workflow as follow:
```yml
name: PR E2E Workflow

on:
  .....
jobs:
  .....
  e2e-node-16:
    name: Electron - Node 16
    # TODO: Change to master branch before the merge
    uses: n8n-io/n8n/.github/workflows/e2e-reusable.yml@parallel_e2e
    with:
      branch: some-branch
      user: oleg
      spec: e2e/*
      run-env: base:16.18.1 # Node 16 with Electron pre-installed
    secrets:
      CYPRESS_RECORD_KEY: ${{ secrets.CYPRESS_RECORD_KEY }}
```

This is important for parallelization because it allows us to keep just a single matrix(containers) that is fully encapsulated within the reusable workflow. 
We hardcode the container matrix to 6 runners but it should be pretty straightforward to calculate this somewhat dynamically. 

## 2. Implement build caching
We use the actions/cache action to ensure Cypress and its dependencies are installed only once and that the tests can begin as soon as possible. We use the specific identifier ${{ github.sha }}-${{ inputs.run-env }}-e2e-modules to take advantage of this action's capabilities. This is very beneficial when re-running the workflow, as it eliminates the need to reinstall the dependencies and rebuild the modules, saving time and resources in the process. Additionally, this helps to ensure that the tests are not impacted by any changes that may have occurred between runs.

For example, here initial run took 13min:
CleanShot 2023-02-10 at 01 22 30@2x
But if we re-run it, it's reduced to 8.5min:
CleanShot 2023-02-10 at 01 32 11@2x

### CODE REVIEW

File: .github/workflows/e2e-reusable.yml
Comment (alexgrozav) [inline .github/workflows/e2e-reusable.yml:87]: Can you set a name here to display the job name as `Spec: "${{ inputs.spec }}"` instead of `testing (x)`?

File: packages/editor-ui/vite.config.ts
Comment (alexgrozav) [inline packages/editor-ui/vite.config.ts:63]: @netroy added this to gain a build speed improvement. Does it need to be removed in order for this to work?

**Review by alexgrozav — DISMISSED** [review]
Beautiful work, @OlegIvaniv!

Approved, but please address comments below before merging.

File: packages/editor-ui/vite.config.ts
Comment (OlegIvaniv) [reply packages/editor-ui/vite.config.ts:63]: Yes, it made the tests that were using Monaco extremely flaky. Since we're now caching builds, this ~2mins save doesn't seem as impactful

Comment (OlegIvaniv) [reply packages/editor-ui/vite.config.ts:63]: I'm going to wait a bit for sure since I'm also running some tests using the Currents dashboard instead of Cypress. But I don't think the PR should block this PR, as the only related part is this small directive

**Review by mutdmour — DISMISSED** [review]

File: packages/cli/src/config/index.ts
Comment (netroy) [inline packages/cli/src/config/index.ts:13]: can we not use `E2E_TESTS` here as well?

**Review by mutdmour — DISMISSED** [review]

**Review by mutdmour — DISMISSED** [review]

**Review by mutdmour — DISMISSED** [review]

File: .github/workflows/e2e-reusable.yml
Comment (mutdmour) [reply .github/workflows/e2e-reusable.yml:87]: not here no.. though there's a github status thing that would show the specs

**Review by mutdmour — DISMISSED** [review]

**Review by mutdmour — DISMISSED** [review]

**Review by mutdmour — DISMISSED** [review]

**Review by mutdmour — DISMISSED** [review]

**Review by mutdmour — DISMISSED** [review]

**Review by mutdmour — DISMISSED** [review]

**Review by mutdmour — DISMISSED** [review]

File: .github/workflows/ci-pull-requests.yml
Comment (mutdmour) [inline .github/workflows/ci-pull-requests.yml:22]: just an opinion but might be better if unit tests and lints happened before smoke test since they are so much slower.

File: .github/workflows/ci-pull-requests.yml
Comment (mutdmour) [inline .github/workflows/ci-pull-requests.yml:1]: for clarity
```suggestion
name: Build, unit/smoke test and lint branch
```

File: .github/workflows/ci-pull-requests.yml
Comment (mutdmour) [inline .github/workflows/ci-pull-requests.yml:106]: for clarity
```suggestion
  smoke-test:
```

File: .github/workflows/e2e-tests-pr.yml
Comment (mutdmour) [inline .github/workflows/e2e-tests-pr.yml:12]: maybe we should stop doing this.. otherwise cypress run hangs and our dashboard looks weird

File: .github/workflows/e2e-reusable.yml
Comment (mutdmour) [inline .github/workflows/e2e-reusable.yml:1]: love this reusable logic here

File: .github/workflows/e2e-reusable.yml
Comment (mutdmour) [inline .github/workflows/e2e-reusable.yml:155]: so it was this simple in the end 😅

File: cypress/e2e/14-data-transformation-expressions.cy.ts
Comment (mutdmour) [inline cypress/e2e/14-data-transformation-expressions.cy.ts:19]: this was intentional.. maybe the name needs to be more unique here and a comment https://github.com/n8n-io/n8n/blob/master/packages/editor-ui/src/views/NodeView.vue#L2597
goal was to block the browser redirect prevention which breaks tests running in headed mode

Comment (OlegIvaniv) [reply cypress/e2e/14-data-transformation-expressions.cy.ts:19]: Yy, I didn't remove it, only fixed the capitalization typo. The event should be `onbeforeunload` not `onBeforeUnload`

Comment (OlegIvaniv) [reply cypress/e2e/14-data-transformation-expressions.cy.ts:19]: Oh, I didn't realize it was our method. Reverted.

File: .github/workflows/e2e-tests-pr.yml
Comment (OlegIvaniv) [reply .github/workflows/e2e-tests-pr.yml]: Commented it out for now

**Review by mutdmour — APPROVED** [review]

**Comment by janober** [comment]
Got released with [`n8n@0.218.0`](https://github.com/n8n-io/n8n/releases/tag/n8n%400.218.0)

**Pull request merged**
